Vertical RTL Book Reading Support

mwgabby-li opened the pull request:
Added a variety of functions to viewport to allow working based on block and inline direction rather than X and Y.

Changed paged mode and flow mode to be based on block and inline directions, making them agnostic to writing direction.

Added jump_to_cfi function to flow mode code, and use this in iframe. This fixes some issues with CFI jumping in flow mode.

Use self.jump_to_cfi in iframe so that it is based on the current mode.

Removed a redundant self.onscroll call that was causing CFIs to be queried twice on load, doubling the time they took.

Fixed some bugs related to scrolling in flow mode related to a decorator not working with keyword arguments, by removing the keyword arguments.

Fixed some bugs with flow mode visibility anchor function.

CFI changes:
 Renamed some functions and variables to clarify meaning.

 Use getBoundingClientRect in range_has_point since it doesn't care about which particular rect matches.

 Remove use of exceptions, since they were causing CFI calculation to run about six times more slowly in benchmarks. (Worst-case went from ~6 to ~1 second.)

 Remove forward flag and allow text offset to be equal to node length, and correct for this by adding a flag to the decoded CFI to indicate this.

 Added comments clarifying the results of decoding a CFI.

 Fix bugs in the point (now decode_with_range) function and simplified it.

 Added new decoded_with_range_to_document_position function, to replace two places where this was being done in slightly different ways.

 scroll_to did not work in my tests, so I fixed the bugs and simplified the logic.

 Rewrote at_current to do a much simplified scan of the viewport in the inline and then block direction.

mwgabby-li commented on src/calibre/ebooks/epub/cfi/parse.py:
```diff
@@ -57,6 +57,9 @@ def __init__(self):
     def parse_epubcfi(self, raw):
         ' Parse a full epubcfi of the form epubcfi(path [ , path , path ]) '
         null = {}, {}, {}, raw
+        if not raw:
+            return null
+
```
This just eliminates an error message, since if raw is not valid an exception will be thrown.

mwgabby-li commented on src/pyj/read_book/cfi.pyj:
```diff
@@ -536,153 +543,155 @@ def at(x, y, doc):  # {{{
+                offset -= 1
+                position_at_end_of_range = True
+
+            range_.setStart(node, offset)
+            range_.setEnd(node, offset + 1)
```
This is important. The "if not offset" code was causing a collapsed range to be created, which doesn't have a valid bounding box. I changed it to always create a range of length 1 starting at the offset (or the offset minus one in the case the offset is equal to the length, with the flag to indicate this.)

mwgabby-li commented on src/pyj/read_book/cfi.pyj:
```diff
@@ -694,74 +703,108 @@ def at_point(ox, oy):  # {{{
+    wini -= 1
+    winb -= 1
+
+    # Don't let the deltas go below 10 or above 30, to prevent perverse cases where
```
That should say "5." Worst-case where the entire viewport is scanned, with this set at the minimum of 5, and there is no CFI found was about 1.5 seconds in my tests.

kovidgoyal commented:
I notice you have code to deal with EPUB cfis being zero/one character over length. Note that this code does not care about the EPUB CFI spec, it only has to deal with CFIs it generates itself, not arbitrary ones from elsewhere. That said I dont object in principle to making it more robust, provided the performance is not significantly impacted.

mwgabby-li commented:
As far as the equal length comment, that would have been better as a meta comment about the change rather than in the code. I'm using the spec as a guideline for consistency's sake. The length part should have no impact on performance.

In fact, the current code is faster because it tends to find valid CFI nodes with fewer calls. In the vast majority of cases, the very first query it makes will find one. In the vast majority of the remaining cases, it will find one after the first two or three outer loops.

As far as the zero offsets: the current code does often generate CFIs with zero offsets, which create collapsed ranges that have no bounding boxes in the version of webkit Calibre is using.
That didn't really matter because before this change, flow mode scrolling to CFIs wasn't actually using the calculations being done in `scroll_to` in cfi.pyj for character offset CFIs.
In iframe, the scroll_to for flow mode CFIs was using the `scroll_to` function directly from cfi.pyj, which means all the calculations in cfi.pyj:601-655 (from `if point_.range is not None:` to `callback(x, y)`) wasn't really being used, because there was no callback sent to this function in flow mode.  So, why did scrolling to CFIs work at all in flow mode?

The reason it worked was because of the `scroll_viewport.scroll_into_view(span)` call on line 609. For spacial offset CFIs, the behavior was better (the else clause, lines 656-670) because of the `scroll_viewport.scroll_to(x, y)` in that function.

Let me know if you have any questions about this code. If you find any issues, let me know and I can look into them.

kovidgoyal commented:
On Sat, Aug 08, 2020 at 10:37:24AM -0700, mwgabby-li wrote:
> In fact, the current code is faster because it tends to find valid CFI nodes with fewer calls. In the vast majority of cases, the very first query it makes will find one. In the vast majority of the remaining cases, it will find one after the first two or three outer loops.
> 
> As far as the zero offsets: the current code does often generate CFIs with zero offsets, which create collapsed ranges that have no bounding boxes in the version of webkit Calibre is using.

Collapsed ranges definitely do have bounding boxes. They would be zero
width/height but they will have valid x, y values. For an example of
that see the code in range_extents() in select.pyj which is used to
position the selection handles while creating highlights. 

> That didn't really matter because before this change, flow mode scrolling to CFIs wasn't actually using the calculations being done in `scroll_to` in cfi.pyj for character offset CFIs.
> In iframe, the scroll_to for flow mode CFIs was using the `scroll_to` function directly from cfi.pyj, which means all the calculations in cfi.pyj:601-655 (from `if point_.range is not None:` to `callback(x, y)`) wasn't really being used, because there was no callback sent to this function in flow mode.  So, why did scrolling to CFIs work at all in flow mode?
> 
> The reason it worked was because of the `scroll_viewport.scroll_into_view(span)` call on line 609. For spacial offset CFIs, the behavior was better (the else clause, lines 656-670) because of the `scroll_viewport.scroll_to(x, y)` in that function.

In fact without scrollIntoView() it wont work at all, that is needed,
because otherwise the position could be inside a div with a scrollbar
and be currently scrolled offscreen. So you cant really get rid of it. 

> 
> Let me know if you have any questions about this code. If you find any issues, let me know and I can look into them.

I will review it in detail when I have time.

mwgabby-li commented:
> Collapsed ranges definitely do have bounding boxes. They would be zero width/height but they will have valid x, y values. For an example of that see the code in range_extents() in select.pyj which is used to position the selection handles while creating highlights.

I thought I was losing my mind, because I had definitely seen this while debugging. I finally figured it out. When you try to get the bounding box of a collapsed range _of vertical text_ with `getClientRects`, the result is always zero. For some reason `getBoundingClientRect` _does_ work though.

The positioning will be slightly more accurate with a non-collapsed range, though, because you can get the inline start and end of the CFI, which is necessary if you want to support CFI side biases correctly. I know that Calibre only uses its own CFIs, so this isn't really a big deal. However, overall, I don't see any disadvantage of using 1-sized ranges, and I'd advocate for keeping them if you don't mind. That said, my priority is vertical reading support in Calibre, and the CFI logic will probably be "good enough" with collapsed ranges.

The only reason I tried to make the CFI code as accurate as possible was to increase the chances of you accepting the pull request. In hindsight, I probably changed more than I should have.

>In fact without scrollIntoView() it wont work at all, that is needed, because otherwise the position could be inside a div with a scrollbar and be currently scrolled offscreen. So you cant really get rid of it.

The problem I was having with this logic is the wrapping of the range in a span. Even after the span was removed, recreating the range didn't seem to actually be working in my tests, and it seemed like I couldn't use the range itself as a scroll target.
This was fine with the normal flow scrolling, because, again, the callback is not bound and the difference between going to the CFI and scrolling to the span is not really noticeable.
I will revisit this and see if I can create something closer to the original behavior. Question: Do you think that maybe I could use `scrollIntoView` on the node the CFI points to instead of the span?
Also, do you have a test book that has divs like this? If so, I would be grateful if you would share it so I can make sure through testing that I can retain the existing behavior with my changes.*

>I will review it in detail when I have time.

I'm especially curious if you find the modified `at_current` function acceptable. I definitely need to modify it to do block and inline so that it scans by line. The function in my pull request scans in the directions of lines, like this, with `B` marking the beginning, `E` the ending, and `o` the end of a line. (Though it will keep scrolling to the end of the page until it finds a CFI, it doesn't really care where the line endings are.)

```
      LTR           Vertical, RTL          RTL
 (i.e. English)    (i.e. Japanese)    (i.e. Hebrew)
| B → → → → → |    | ↓ ↓ ↓ ↓ ↓ B |   | ← ← ← ← ← B |
| → → → → o   |    | ↓ ↓ o ↓ ↓ ↓ |   |   o ← ← ← ← |
| → → → → → → |    | ↓ ↓   ↓ ↓ ↓ |   | ← ← ← ← ← ← |
| → → o       |    | ↓ ↓   ↓ o ↓ |   |       o ← ← |
| → → → → → → |    | E ↓   ↓   ↓ |   | ← ← ← ← ← ← |
| → → → o     |    |   ↓   ↓   ↓ |   |     o ← ← ← |
| → → → → → E |    |   ↓   ↓   ↓ |   | E ← ← ← ← ← |
```
I'm pretty happy with it, and I went to a lot of trouble making sure it found the first thing at the start of the block and inline direction. I also was thinking that it might be better for cross-compatibility between bookmarks in flow mode and paged mode if this sort of CFI search was used in both modes. Currently, paged mode tries to find a CFI in the center of the viewport instead of scanning.

*The div issue is a larger one, in my opinion, and opens up other questions. What if you're reading a div, but the CFI found is not inside the div? You could solve that problem by having the user select something and have that be used as a CFI for a bookmark; or maybe you could cache the user's last touch or click or scroll position as a first query for a CFI if it was inside a scrolling div. Anyway, I'm not trying to solve these issues, but I thought I'd bring it up.

kovidgoyal commented:
On Sat, Aug 08, 2020 at 04:58:15PM -0700, mwgabby-li wrote:
> > Collapsed ranges definitely do have bounding boxes. They would be zero width/height but they will have valid x, y values. For an example of that see the code in range_extents() in select.pyj which is used to position the selection handles while creating highlights.
> 
> I thought I was losing my mind, because I had definitely seen this while debugging. I finally figured it out. When you try to get the bounding box of a collapsed range _of vertical text_ with `getClientRects`, the result is always zero. For some reason `getBoundingClientRect` _does_ work though.
> 
> The positioning will be slightly more accurate with a non-collapsed range, though, because you can get the inline start and end of the CFI, which is necessary if you want to support CFI side biases correctly. I know that Calibre only uses its own CFIs, so this isn't really a big deal. However, overall, I don't see any disadvantage of using 1-sized ranges, and I'd advocate for keeping them if you don't mind. That said, my priority is vertical reading support in Calibre, and the CFI logic will probably be "good enough" with collapsed ranges.
> 
> The only reason I tried to make the CFI code as accurate as possible was to increase the chances of you accepting the pull request. In hindsight, I probably changed more than I should have.

I am OK with moving to 1 based ranges in principle, provided there are
no performance/accuracy regressions. But, you are correct that all the
changes to cfi.pyj make it much more work to review. 

> 
> >In fact without scrollIntoView() it wont work at all, that is needed, because otherwise the position could be inside a div with a scrollbar and be currently scrolled offscreen. So you cant really get rid of it.
> 
> The problem I was having with this logic is the wrapping of the range in a span. Even after the span was removed, recreating the range didn't seem to actually be working in my tests, and it seemed like I couldn't use the range itself as a scroll target.

Why would recreating the range fail? You just store the start and end
nodes and offsets from before inserting the span then use
setBaseAndExtents on a new range with the saved values.

> This was fine with the normal flow scrolling, because, again, the callback is not bound and the difference between going to the CFI and scrolling to the span is not really noticeable.
> I will revisit this and see if I can create something closer to the original behavior. Question: Do you think that maybe I could use `scrollIntoView` on the node the CFI points to instead of the span?

No, that wont work, think of a paragraph with lots of text,
scrollIntoView will only ensure the top of the paragraph is visible, the
actual position might be lower down and require more scrolling. This can
be worked around by doing manual scrolling in addition to scroll into
view but just using scrollintoview is more robust and elegant, since you
have to use it anyway.


> Also, do you have a test book that has divs like this? If so, I would be grateful if you would share it so I can make sure through testing that I can retain the existing behavior with my changes.*

Just create use the following markup:

<div style="width: 100%; height: 200px">
... lots of random text ...
</div>

> 
> >I will review it in detail when I have time.
> 
> I'm especially curious if you find the modified `at_current` function acceptable. I definitely need to modify it to do block and inline so that it scans by line. The function in my pull request scans in the directions of lines, like this, with `B` marking the beginning, `E` the ending, and `o` the end of a line. (Though it will keep scrolling to the end of the page until it finds a CFI, it doesn't really care where the line endings are.)
> 
> ```
>       LTR           Vertical, RTL          RTL
>  (i.e. English)    (i.e. Japanese)    (i.e. Hebrew)
> | B → → → → → |    | ↓ ↓ ↓ ↓ ↓ B |   | ← ← ← ← ← B |
> | → → → → o   |    | ↓ ↓ o ↓ ↓ ↓ |   |   o ← ← ← ← |
> | → → → → → → |    | ↓ ↓   ↓ ↓ ↓ |   | ← ← ← ← ← ← |
> | → → o       |    | ↓ ↓   ↓ o ↓ |   |       o ← ← |
> | → → → → → → |    | E ↓   ↓   ↓ |   | ← ← ← ← ← ← |
> | → → → o     |    |   ↓   ↓   ↓ |   |     o ← ← ← |
> | → → → → → E |    |   ↓   ↓   ↓ |   | E ← ← ← ← ← |
> ```
> I'm pretty happy with it, and I went to a lot of trouble making sure it found the first thing at the start of the block and inline direction. I also was thinking that it might be better for cross-compatibility between bookmarks in flow mode and paged mode if this sort of CFI search was used in both modes. Currently, paged mode tries to find a CFI in the center of the viewport instead of scanning.

The reason for paged mode doing that is that it is more likely to give a
better result for preserving page position when the viewport is resized.
After a resize all pages are different, so the top of the page could be
the bottom of the new page and restoring to that would look like the
viewer has restored to the following page. Using the middle of the page 
means that at least some of the original page is likely to be present on
the restored page. 
-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

mwgabby-li commented:
OK, I found the issue. It seems that recreating from the saved nodes doesn't work because they've been invalidated by `surroundContents`. Since the CFI ranges used in the scroll to logic only contain one node, using the parent of the span seems safe? That should be the original line of text, I think? Anyway, when I tested this, I got exactly the same coordinates back from using the parent node range that I did from the original range.

>Using the middle of the page means that at least some of the original page is likely to be present on the restored page.

Well, one workaround would be to make two CFI bookmarks in paged mode, and use one as a 'flow mode' bookmark and the other as a 'paged mode' bookmark. Just brainstorming!

kovidgoyal commented:
On Sat, Aug 08, 2020 at 09:47:06PM -0700, mwgabby-li wrote:
> OK, I found the issue. It seems that recreating from the saved nodes doesn't work because they've been invalidated by `surroundContents`. Since the CFI ranges used in the scroll to logic only contain one node, using the parent of the span seems safe? That should be the original line of text, I think? Anyway, when I tested this, I got exactly the same coordinates back from using the parent node range that I did from the original range.
> 

If they are text nodes then yes they will be invalidated. In such a case
you can check if its a text node, and store the parent + index into
childNodes for that node + offset. Remember to call normalize() before
saving and before restoring.

> >Using the middle of the page means that at least some of the original page is likely to be present on the restored page.
> 
> Well, one workaround would be to make two CFI bookmarks in paged mode, and use one as a 'flow mode' bookmark and the other as a 'paged mode' bookmark. Just brainstorming!

Yes, that is a possibility, specifically for bookmarks, but its rather a
lot of work/changes for a somewhat obscure issue. A bookmark restoring
to "approximate" location is typically good enough.

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

mwgabby-li commented:
Yes, it would always be a text node in this codepath, I believe? This is the CFI text offset path. I've updated my PR with your suggestions, and made `scroll_to` behave much more like the original code. I hope this makes it easier to review and addresses your concerns so far. 

The original code (the code I didn't write) was saving the range's nodes before creating the span, and trying to restore the range from those.

The span is being removed, the nodes are being added, and the parent is being normalized like you are saying in the original code. In my change, all I did was send the parent node to the range with the saved offsets instead of the invalid saved nodes. I changed it so it doesn't save them anymore, just the offsets.

There was a test failure on windows-2016, but it doesn't look related to what I did. If you need me to try to run the test and fix it, please let me know. 

On Sat, Aug 8, 2020, at 22:44, Kovid Goyal wrote:
> 

> On Sat, Aug 08, 2020 at 09:47:06PM -0700, mwgabby-li wrote:
> > OK, I found the issue. It seems that recreating from the saved nodes doesn't work because they've been invalidated by `surroundContents`. Since the CFI ranges used in the scroll to logic only contain one node, using the parent of the span seems safe? That should be the original line of text, I think? Anyway, when I tested this, I got exactly the same coordinates back from using the parent node range that I did from the original range.
> > 
> 
> If they are text nodes then yes they will be invalidated. In such a case
> you can check if its a text node, and store the parent + index into
> childNodes for that node + offset. Remember to call normalize() before
> saving and before restoring.
> 
> > >Using the middle of the page means that at least some of the original page is likely to be present on the restored page.
> > 
> > Well, one workaround would be to make two CFI bookmarks in paged mode, and use one as a 'flow mode' bookmark and the other as a 'paged mode' bookmark. Just brainstorming!
> 
> Yes, that is a possibility, specifically for bookmarks, but its rather a
> lot of work/changes for a somewhat obscure issue. A bookmark restoring
> to "approximate" location is typically good enough.
> 
> -- 
> _____________________________________
> 
> Dr. Kovid Goyal 
> https://www.kovidgoyal.net
> https://calibre-ebook.com
> _____________________________________
> 

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/kovidgoyal/calibre/pull/1206#issuecomment-671009941>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AANTD44WGS7P45BD6D62QSLR7YZU7ANCNFSM4PYVECCQ>.

kovidgoyal commented:
Ignore the test failure, that's just the test machine flaking. I'll
review when I have a moment.

On Sat, Aug 08, 2020 at 11:07:43PM -0700, mwgabby-li wrote:
> Yes, it would always be a text node in this codepath, I believe? This is the CFI text offset path. I've updated my PR with your suggestions, and made `scroll_to` behave much more like the original code. I hope this makes it easier to review and addresses your concerns so far. 
> 
> The original code (the code I didn't write) was saving the range's nodes before creating the span, and trying to restore the range from those.
> 
> The span is being removed, the nodes are being added, and the parent is being normalized like you are saying in the original code. In my change, all I did was send the parent node to the range with the saved offsets instead of the invalid saved nodes. I changed it so it doesn't save them anymore, just the offsets.
> 
> There was a test failure on windows-2016, but it doesn't look related to what I did. If you need me to try to run the test and fix it, please let me know. 
> 
> On Sat, Aug 8, 2020, at 22:44, Kovid Goyal wrote:
> > 
> 
> > On Sat, Aug 08, 2020 at 09:47:06PM -0700, mwgabby-li wrote:
> > > OK, I found the issue. It seems that recreating from the saved nodes doesn't work because they've been invalidated by `surroundContents`. Since the CFI ranges used in the scroll to logic only contain one node, using the parent of the span seems safe? That should be the original line of text, I think? Anyway, when I tested this, I got exactly the same coordinates back from using the parent node range that I did from the original range.
> > > 
> > 
> > If they are text nodes then yes they will be invalidated. In such a case
> > you can check if its a text node, and store the parent + index into
> > childNodes for that node + offset. Remember to call normalize() before
> > saving and before restoring.
> > 
> > > >Using the middle of the page means that at least some of the original page is likely to be present on the restored page.
> > > 
> > > Well, one workaround would be to make two CFI bookmarks in paged mode, and use one as a 'flow mode' bookmark and the other as a 'paged mode' bookmark. Just brainstorming!
> > 
> > Yes, that is a possibility, specifically for bookmarks, but its rather a
> > lot of work/changes for a somewhat obscure issue. A bookmark restoring
> > to "approximate" location is typically good enough.
> > 
> > -- 
> > _____________________________________
> > 
> > Dr. Kovid Goyal 
> > https://www.kovidgoyal.net
> > https://calibre-ebook.com
> > _____________________________________
> > 
> 
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub <https://github.com/kovidgoyal/calibre/pull/1206#issuecomment-671009941>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AANTD44WGS7P45BD6D62QSLR7YZU7ANCNFSM4PYVECCQ>.
> 
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/kovidgoyal/calibre/pull/1206#issuecomment-671011545

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

kovidgoyal commented:
> > Collapsed ranges definitely do have bounding boxes. They would be zero width/height but they will have valid x, y values. For an example of that see the code in range_extents() in select.pyj which is used to position the selection handles while creating highlights.
> 
> I thought I was losing my mind, because I had definitely seen this while debugging. I finally figured it out. When you try to get the bounding box of a collapsed range _of vertical text_ with `getClientRects`, the result is always an empty list of bounding boxes. For some reason `getBoundingClientRect` _does_ work though.
> 

So, you cant use getBoundingClientRect in paged mode. This is because if the range stretches across two css columns, it will be the union of clientrects in the two columns and therefore be meaningless. I see you have changed use of clientRects with boundingrect in a  few places, for instance, in range_has_point(). As I understand it, range_has_point is not called with empty ranges, so why the change?

mwgabby-li commented:
>I see you have changed use of clientRects with boundingrect in a few places, for instance, in range_has_point().

It's not necessary. I thought it was an optimization and a simplification. I will remove it from `range_has_point`.

However, in the cases of a range created from a CFI, I don't think it's a problem because such a range will only be one character, which should not be able to go across multiple columns? It simplifies the code in such cases not to bother having a list. If you prefer, I will remove it throughout cfi.pyj except in cases where it was there in the original code.

kovidgoyal commented:
On Sun, Aug 09, 2020 at 09:17:08PM -0700, mwgabby-li wrote:
> >I see you have changed use of clientRects with boundingrect in a few places, for instance, in range_has_point().
> 
> It's not necessary. I thought it was an optimization and a simplification. I will remove it from `range_has_point`.

Sounds good.

> 
> However, in the cases of a range created from a CFI, I don't think it's a problem because such a range will only be one character, which should not be able to go across multiple columns? It simplifies the code in such cases not to bother having a list. If you prefer, I will remove it throughout cfi.pyj except in cases where it was there in the original code.

Assuming that the range is a single character in a text node, then yes,
it is not a problem.

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

mwgabby-li commented:
When testing the `getBoundingClientRect` to `getClientRects` change, I discovered that using `p` in cfi.pyj's `scroll_to` to restore from doesn't make sense, because this is the parent of the text node, not the text node itself. I have no idea why it sometimes worked in my tests!

I changed the code to save the node's original parent and its offset in the parent, and use that to get the children.
Does this make sense to you? I'm not a DOM expert like you, I'm totally a beginner!

Do you think it's better to use the saved parent, like so:

```
# Save original parent
node_parent = original_node.parentNode
original_node_index = 0
for child_node in node_parent.childNodes:
    if child_node is original_node:
        break
    else:
        ++original_node_index
<...>
fn = def():
    # Remove the span we inserted
    p = span.parentNode
    for node in span.childNodes:
        span.removeChild(node)
        p.insertBefore(node, span)
    p.removeChild(span)
    p.normalize()

    # Reset the range to what it was before the span was added
    offset = so
    while offset > -1:
        try:
            # Restore from original parent, node_parent:
            decoded.range.setStart(node_parent.childNodes[original_node_index], offset)
            break
        except:
            offset -= 1
```
or this?

```
            # Restore from normalized node, p:
            decoded.range.setStart(p.childNodes[original_node_index], offset)
```

kovidgoyal commented:
unwrapping the span can *only* affect the *contents* of parent node, never the parent node itself, since it calls only methods that act on the contents of the parent node. Therefore it should be perfectly safe to store a reference to the parent node along with the index of the text node in it.

mwgabby-li commented:
Thanks! I've push a fix for the feedback so far, as well as a crash I found while testing the feedback.

When doing a diff of cfi.pyj against the one in the original py3 branch, besides `range_has_point`, I didn't see any cases where the original code used `getClientRects` on a range longer than one character, so that's the only thing I changed back. Please let me know if I missed anything.

kovidgoyal commented:
You need to rebase your branch on py3 then force push it

mwgabby-li commented:
Sorry, I'm still learning git, and I broke something. I'm going to close this pull request for now. Let me know if and when you have time to look at it and I'll fork the repository and put it up again.

To make it easier to merge, I can try to split it into smaller pieces:
The CFI code and the viewport changes could be one thing.
The writing-mode fixups is completely independent.
The paged mode changes and flow mode changes are dependent on the other two things, but could be done third.

kovidgoyal commented:
You dont need to close it, just do a 

git rebase origin/py3
git push -f

mwgabby-li commented:
Ouch, sorry. I didn't see your comment before I deleted the repo again.

kovidgoyal commented:
Well go ahead and open a new PR I will look at it in the next few days.
